### PR TITLE
[BUG] fix value mask bug in dp_critic

### PR DESCRIPTION
**Pull request by Jackory**
### Checklist Before Starting

- [x] Search for similar PR(s).

### What does this PR do?

> Add one-line overview of what this PR aims to achieve or accomplish. 

When critic.use_dynamic_size is enabled, values rearrange indices but attention_mask does not, causing values * attention_mask to produce unpredictable bugs. This bug may have affected nearly all previous PPO-based experiments if critic.use_dynamic_size was turned on.

### High-Level Design

> Demonstrate the high-level design if this PR is complex.

### Specific Changes

> List the specific changes.

### API

> Demonstrate how the API changes if any.

### Usage Example

> Provide usage example(s) for easier usage.

```python
# Add code snippet or script demonstrating how to use this 
```

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluatuion results, etc.

### Additional Info.

- **Issue Number**: Fixes issue # or discussion # if any.
- **Training**: [Note which backend this PR will affect: FSDP, Megatron, both, or none]
- **Inference**: [Note which backend this PR will affect: vLLM, SGLang, both, or none]

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl?tab=readme-ov-file#contribution-guide).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl?tab=readme-ov-file#code-linting-and-formatting).
- [x] Add `[BREAKING]` to the PR title if it breaks any API.
- [x] Update the documentation about your changes in the [docs](https://github.com/volcengine/verl/tree/main/docs).
- [x] Add CI test(s) if neccessary.

### CODE REVIEW

**Review by vermouth1992 — APPROVED** [review]

**Comment by vermouth1992** [comment]
Great catch! Thank you!

**Comment by vermouth1992** [comment]
@tongyx361 Could you please confirm the changes?

File: verl/workers/critic/dp_critic.py [hunk 1]
```diff
@@ -160,14 +160,13 @@ def compute_values(self, data: DataProto) -> torch.Tensor:
...
         response_length = responses.size(1)
-        values = values * attention_mask[:, -response_length - 1 : -1]
 
         if use_dynamic_bsz:
             indices = list(itertools.chain.from_iterable(indices))
             assert len(indices) == values.size(0), f"{len(indices)} vs. {values.size()}"
             revert_indices = torch.tensor(get_reverse_idx(indices), dtype=torch.long)
             values = values[revert_indices]
-
+        values = values * attention_mask[:, -response_length - 1 : -1]
```
Comment (tongyx361) [inline verl/workers/critic/dp_critic.py:169]: This also looks correct to me! What about moving the reverting logic immediately after the `values` computation to make the `use_dynamic_bsz` more local?

```suggestion
        # Revert the `values`' order back after rearranging to keep consistent with original `data`
        if use_dynamic_bsz:
            indices = list(itertools.chain.from_iterable(indices))
            assert len(indices) == values.size(0), f"{len(indices)} vs. {values.size()}"
            revert_indices = torch.tensor(get_reverse_idx(indices), dtype=torch.long)
            values = values[revert_indices]
        responses = data.batch["responses"]
        attention_mask = data.batch["attention_mask"]
        response_length = responses.size(1)
        values = values * attention_mask[:, -response_length - 1 : -1]
```

**Comment by eric-haibin-lin** [comment]
@ETOgaosion could u help check if mcore have the same issue?
